ci-operator: ask resolver for config with additional test inject

petr-muller opened the pull request:
/hold

Builds over https://github.com/openshift/ci-tools/pull/2198

Code is a bit hacky, but I wanted to have something working. Probably needs tests and maybe an integration test. I did not want to change `load.Config` signature, so I chose to use functional option pattern. An annoying part of the whole thing is that `config.Load` is able to read the config from million different sources (files envvar, resolved, unresolved etc), but the injection only works with one of them, which may be unintuitive. /shrug

stevekuznetsov commented on pkg/load/load.go:
```diff
@@ -121,10 +122,57 @@ func fromPath(path string) (filenameToConfig, error) {
-func Config(path, unresolvedPath, registryPath string, info *ResolverInfo) (*api.ReleaseBuildConfiguration, error) {
+var testPathRegex = regexp.MustCompile(`(?P<org>[^/]+)/(?P<repo>[^@]+)@(?P<branch>[^:]+):(?P<test>.+)`)
+
+func parseCiopTestPath(path string) (ciopTestPath, error) {
```
Can we just pass JSON please to the flag

petr-muller replied:
you mean `--inject-test='{"org": "o","repo": "r","branch": "b","test": "t"}`? That's super ugly and terrible to use...

I know you hate parsing information from strings, but string inputs are comfortable and readable, and regex parsing is not even that ugly.

petr-muller replied:
I don't expect humans to pass this, but I expect humans (mostly DPTP) to occasionally eyeball the calls with the params, and I think it is useful if they are readable.

I don't find regex parsing here that complex or fragile - it's not using any complicated patterns that basically only deals with separators) and it could be even read by some `scanf` (using regex with named groups instead of scanf is probably my Python-influenced preference). Of course I'd add tests.

How strong are you on this? I agree that parsing JSON will make the code simpler, but at the expense of human observability, which is IMO rarely a good trade.

stevekuznetsov replied:
Humans should not read the flags we generate on the fly inside of a ProwJob YAML either - let's make the `ci-operator` log output clear when it's doing this on resolution?

stevekuznetsov commented on pkg/load/load.go:
```diff
@@ -186,22 +234,42 @@ func Config(path, unresolvedPath, registryPath string, info *ResolverInfo) (*api
 }
 
-func configFromResolver(info *ResolverInfo) (*api.ReleaseBuildConfiguration, error) {
+func configFromResolver(info *ResolverInfo, options configLoadOpt) (*api.ReleaseBuildConfiguration, error) {
```
`ResolverInfo` is already a structure to hold the data we use to query the resolver - why not extend it?

petr-muller replied:
My intent is to explore the dynamic injection concept further in the future, I want to try things like "inject a workflow test". That leads to different resolver endpoints with different sets of parameters, and I don't think we want to have a single struct that's basially an union of all different parameters?

I'll try to invest a bit more effort into refactoring the client code, let's see if I can try to come up with something better.

stevekuznetsov commented on pkg/load/load.go:
```diff
@@ -186,22 +234,42 @@ func Config(path, unresolvedPath, registryPath string, info *ResolverInfo) (*api
-func configFromResolver(info *ResolverInfo) (*api.ReleaseBuildConfiguration, error) {
+func configFromResolver(info *ResolverInfo, options configLoadOpt) (*api.ReleaseBuildConfiguration, error) {
+	queryParams := map[string]string{
+		"org":    info.Org,
```
surely we have constants for those query parameters?

petr-muller replied:
I guess, but the code here did not use them.

stevekuznetsov commented on pkg/load/load.go:
```diff
@@ -186,22 +234,42 @@ func Config(path, unresolvedPath, registryPath string, info *ResolverInfo) (*api
+	}
 	identifier := fmt.Sprintf("%s/%s@%s", info.Org, info.Repo, info.Branch)
 	if info.Variant != "" {
 		identifier = fmt.Sprintf("%s [%s]", identifier, info.Variant)
```
Isn't this logic (and the version on the injected one) already a method on `api.Metadata` ? Why doesn't the `ResolverInfo` embed one?

petr-muller replied:
But it doesn't seem this logic is a method on api.Metadata

petr-muller commented:
@stevekuznetsov given that your feedback is more about the existing code and not my changes, let's hold this and I'll try to refactor the code to something more sensible.

petr-muller commented:
Extracted the existing code refactors to separate PRs:

- https://github.com/openshift/ci-tools/pull/2245
- https://github.com/openshift/ci-tools/pull/2246

stevekuznetsov commented:
Will review after rebase

petr-muller commented:
@stevekuznetsov PTAL, took a slightly different approach than previous version (`ResolverClient` and using its new method directly from ci-op, instead of making `load.Config` more complicated). 

For input feedback, I gave some arguments in https://github.com/openshift/ci-tools/pull/2215#discussion_r693500683 and wrote some tests to deal with potential fragility, but it's not an issue I'd like to get blocked on, so if you feel strongly about it, let me know and I'll just let it parse JSON.

petr-muller commented:
/hold cancel

stevekuznetsov commented on pkg/api/metadata.go:
```diff
@@ -134,3 +134,28 @@ func BuildCacheFor(metadata Metadata) ImageStreamTagReference {
+	ret.Repo = match[2]
+	ret.Test = match[4]
+
+	branchAndVariant := strings.Split(match[3], "__")
```
Branches can contain `__` - not sure it matters to handle it, but also why IMO if you're going to stringify some data it's better to do it to a spec.

petr-muller replied:
It's the same specification we use in ci-op naming convention though, so afaik we'll break there already for branches that contain `__`.

stevekuznetsov commented on pkg/registry/server/client.go:
```diff
@@ -0,0 +1,118 @@
+package server
```
`client.go` in `package server`? lol

petr-muller replied:
Honestly our package game is :hankey: :rofl: 

Any suggestions? I think I previously had `pkg/registry/client/client.go` but then decided to merge them together b/c the client is coupled with the server via param names etc.

stevekuznetsov commented on pkg/registry/server/client.go:
```diff
@@ -0,0 +1,118 @@
+		return nil, fmt.Errorf("failed to create request for configresolver: %w", err)
+	}
+	query := req.URL.Query()
+	query.Add(InjectTestQuery, testSource.Test)
```
```go
for query, value := range map[string]string {
    InjectTestQuery: testSource.Test,
// ...
} {
    query.add(query, value)
}
```

?

petr-muller replied:
I considered that when I wrote it, but then I thought it has no advantages in this specific case. It just makes code more indirect. You'd still need code lines to make that map.

petr-muller replied:
Addressed

stevekuznetsov commented on pkg/registry/server/client_test.go:
```diff
@@ -0,0 +1,326 @@
+	"github.com/openshift/ci-tools/pkg/api"
+)
+
+var parsedConfig = &api.ReleaseBuildConfiguration{
```
Why is this a global (and not a `testdata/` entry?)

petr-muller replied:
This existed before, I just moved it. I guess I can modernize it.

petr-muller replied:
Addressed

petr-muller commented on pkg/registry/server/client.go:
```diff
@@ -0,0 +1,118 @@
+	query := req.URL.Query()
+	query.Add(InjectTestQuery, testSource.Test)
+	query.Add(OrgQuery, base.Org)
+	query.Add(InjectTestQuery, testSource.Org)
```
bug

petr-muller commented:
/test images
Registry flake: `Error fetching blob: invalid status code from registry 503 (Service Unavailable)`

@stevekuznetsov addressed all feedback

stevekuznetsov commented:
/lgtm
/hold cancel

petr-muller commented:
Thanks for all the input here!
